issue: 4043157 Set initial RTO to 1 second per RFC 6298

tomerdbz opened the pull request:
## Description
Change initial TCP retransmission timeout from 3 seconds to 1 second as recommended by RFC 6298 Section 2.

RFC 6298 states: "Until a round-trip time (RTT) measurement has been made... the sender SHOULD set RTO <- 1 second"

This applies to all new TCP connections until an RTT measurement is made. Once the connection is established and ACKs received, TCP calculates actual RTT and updates RTO accordingly using the Van Jacobson algorithm.

Modified functions:
- tcp_pcb_init(): Set rto/sv to 1000ms (was 3000ms)
- tcp_pcb_recycle(): Set rto/sv to 1000ms (was 3000ms)

This fixes incorrect SYN retransmission timing where the first retry occurred after ~3 seconds instead of ~1 second, causing TCP_USER_TIMEOUT tests to fail and slower connection failure detection.

Benefits:
- TCP/IP standards compliance (RFC 6298)
- Faster connection establishment failure detection
- Matches Linux kernel TCP_TIMEOUT_INIT behavior

##### What
Set initial RTO to 1 second per RFC 6298

##### Why ?
RFC Compliance.

##### How ?
_It is optional but for complex PRs please provide information about the design,
architecture, approach, etc._

## Change type
What kind of change does this PR introduce?
- [X] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Tests
- [ ] Other

## Check list
- [ ] Code follows the style de facto guidelines of this project
- [ ] Comments have been inserted in hard to understand places
- [ ] Documentation has been updated (if necessary)
- [ ] Test has been added (if possible)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Chores**
  * Improved TCP retransmission timeout handling for more reliable connection establishment and faster recovery during packet loss or network congestion scenarios.

<!-- review_stack_entry_start -->

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/Mellanox/libxlio/pull/498)

<!-- review_stack_entry_end -->
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tomerdbz committed:
issue: 4043157 Set initial RTO to 1 second per RFC 6298

Change initial TCP retransmission timeout from 3 seconds to
1 second as recommended by RFC 6298 Section 2, with additional
safeguards for timer granularity and RFC 1122 compliance.

RFC 6298 states: "Until a round-trip time (RTT) measurement
has been made... the sender SHOULD set RTO <- 1 second"

This applies to all new TCP connections until an RTT
measurement is made.

Changes:

1. TCP RTO Calculation (src/core/lwip/tcp.c):
   - Added get_initial_rto() helper function
   - Uses round-up division
   - Prevents division trunc that could cause premature timeouts
   - Modified tcp_pcb_init() and tcp_pcb_recycle() to use helper

2. Timer Resolution Limits (src/core/util/sys_vars.cpp):
   - Added RFC 1122 validation for tcp_timer_resolution_msec
   - Enforces maximum of 500ms per RFC 1122 Section 4.2.3.2
   - Logs warning and clamps value if exceeded
   - Applied to both environment variable and config registry paths

3. Configuration Schema (xlio_config_schema.json):
   - Added "maximum": 500 constraint to timer_msec
   - Updated description to reference RFC 1122 requirement
   - Prevents invalid configurations at schema validation level

4. Documentation (README):
   - Added RFC 1122 reference to timer_msec documentation

Rationale:
Previous implementation used simple division (1000 / slow_tmr_interval)
which could result in:
- 0 ticks when interval > 1000ms (immediate timeout)
- 1 tick when interval = 1000ms (fires on next tick)
- Insufficient granularity with large timer intervals

The new implementation provides defense-in-depth:
- Schema validation prevents misconfiguration
- Runtime validation enforces RFC 1122 limits (delayed ACK ≤ 500ms)
- Round-up division ensures minimum 1 tick without truncation

This fixes incorrect SYN retransmission timing where the
first retry could occur prematurely or immediately, causing
TCP_USER_TIMEOUT tests to fail and connection establishment
issues.

Benefits:
- TCP/IP standards compliance (RFC 6298 and RFC 1122)
- Robust handling of timer granularity edge cases
- Faster connection establishment failure detection
- Prevents problematic timer configurations
- Matches Linux kernel TCP_TIMEOUT_INIT behavior

Signed-off-by: Tomer Cabouly <tcabouly@nvidia.com>

galnoam commented:
@pasis  can you check if we want this fix? Sync with @tomerdbz please.

pasis commented on src/core/lwip/tcp.c:
```diff
@@ -905,6 +905,11 @@ err_t tcp_recv_null(void *arg, struct tcp_pcb *pcb, struct pbuf *p, err_t err)
 
+static inline u32_t get_initial_rto(void)
+{
+    return (1000 + slow_tmr_interval - 1) / slow_tmr_interval;
```
Until we improve timer precision, the number of ticks needs to be > 1. Otherwise, RTO can be triggered immediately mistakenly.

tomerdbz replied:
Fixed

pasis commented on src/core/util/sys_vars.cpp:
```diff
@@ -2744,6 +2754,16 @@ void mce_sys_var::configure_memory_limits(const config_registry &registry)
+                    "Clamping to 500ms to ensure protocol compliance.\n",
+                    SYS_VAR_TCP_TIMER_RESOLUTION_MSEC, tcp_timer_resolution_msec);
+        tcp_timer_resolution_msec = 500;
+    }
```
If we decide to keep the upper limit, we don't need this check because it's in the schema.

[comment] We don't have upper limit for the basic timer, right? In this case if timer > tcp_timer, limiting tcp_timer back to 500ms will calculate ticks based on this value while actual ticks will still run with the basic timer granularity.

tomerdbz replied:
Unofrutantely it's needed for the legacy env vars - but you're right we can remove one instance (for the new config)

tomerdbz replied:
removed this from the PR

tomerdbz committed:
issue: 4043157 Add RFC 6298 SYN backoff and Section 5.7 fallback RTO

Follow-up to the initial RTO change (3s -> 1s). With a sub-3-second
initial RTO, two additional RFC 6298 requirements must be addressed:

1. Section 5.5 - Exponential backoff during SYN_SENT:
   Previously, RTO backoff was skipped entirely in SYN_SENT state,
   causing fixed-interval SYN retransmissions (every 1s). Now applies
   standard exponential backoff: 1s, 2s, 4s, 8s, 16s, 32s -- matching
   Linux kernel behavior.

2. Section 5.7 - Fallback RTO after SYN retransmission (MUST):
   If SYN or SYN-ACK was retransmitted, re-initialize RTO to 3 seconds
   when entering ESTABLISHED state. This prevents spurious data
   retransmissions on paths with RTT > 1s where the initial RTO was
   insufficient. Applied to both active (SYN_SENT) and passive
   (SYN_RCVD) connection establishment.

Also:
- Introduced TCP_INITIAL_RTO_MS and TCP_FALLBACK_RTO_MS named constants
  in opt.h with #ifndef guards for overridability.
- Moved tcp_ms_to_ticks() and tcp_fallback_rto() to tcp_impl.h as
  static inline helpers.
- Skip congestion window reduction during SYN_SENT retransmission
  (no cwnd/ssthresh to reduce before ESTABLISHED).

Signed-off-by: Tomer Cabouly <tcabouly@nvidia.com>

tomerdbz commented:
libvma Relevance Analyzer
────────────────────────────────────────
📁 Analyzing 4 file(s)...
⏳ Connecting to GitHub... done

  [1/4] lwip/opt.h... ⚪ not relevant
  [2/4] lwip/tcp.c (3 hunks)... ✅ relevant
  [3/4] lwip/tcp_impl.h... ⚪ not relevant
  [4/4] lwip/tcp_in.c (2 hunks)... ✅ relevant

────────────────────────────────────────
📋 Summary
────────────────────────────────────────

✅ 2 change(s) MAY BE RELEVANT to libvma:
   • src/core/lwip/tcp.c
     → src/vma/lwip/tcp.c
     libvma has the same feature/logic that was changed in libxlio, which is the retransmission time-out logic in the tcp_slowtmr function. The change in libxlio seems to be a bug fix or improvement, as it follows the RFC 6298 5.5 for exponential backoff of the retransmission timer. This change is relevant to libvma as it can improve the performance and reliability of the TCP implementation.
   • src/core/lwip/tcp_in.c
     → src/vma/lwip/tcp_in.c
     libvma has the same feature/logic related to TCP retransmission and RTO calculation as libxlio, and the change in libxlio appears to be a bug fix or improvement based on RFC 6298, which is also relevant to libvma
     
TODO - add task to libvma

tomerdbz commented on src/core/config/descriptor_providers/xlio_config_schema.json:
```diff
@@ -418,9 +418,10 @@
                                 "timer_msec": {
                                     "type": "integer",
                                     "minimum": 0,
+                                    "maximum": 500,
```
why? this violates RFC

tomerdbz replied:
removed this from the PR

tomerdbz commented on src/core/util/sys_vars.cpp:
```diff
@@ -1621,6 +1621,16 @@ void mce_sys_var::get_env_params()
+                    "Clamping to 500ms to ensure protocol compliance.\n",
+                    SYS_VAR_TCP_TIMER_RESOLUTION_MSEC, tcp_timer_resolution_msec);
+        tcp_timer_resolution_msec = 500;
+    }
```
we can remove the maximum enforcement and just warn - but I still don't understand why

tomerdbz replied:
removed this from the PR

tomerdbz commented on src/core/lwip/tcp_in.c:
```diff
@@ -574,6 +574,14 @@ static err_t tcp_process(struct tcp_pcb *pcb, tcp_in_data *in_data)
+                u32_t fallback = tcp_fallback_rto();
+                pcb->rto = fallback;
+                pcb->sv = fallback;
+            }
```
Agreed - thanks for the thoughtful comment dear AI

tomerdbz commented:
bot:retest

pasis commented on src/core/lwip/tcp_impl.h:
```diff
@@ -320,8 +320,19 @@ extern u32_t rcv_wnd_scale;
+extern u32_t slow_tmr_interval;
 extern ip_route_mtu_fn external_ip_route_mtu;
 
+static inline u32_t tcp_ms_to_ticks(u32_t ms)
```
use this function for all ticks conversions. In tcp.c there are multiple places where the old way is used.

tomerdbz replied:
Fixed

tomerdbz committed:
issue: 4043157 Use tcp_ms_to_ticks() for all timer conversions in tcp.c

Replace 8 hand-rolled "ms / slow_tmr_interval" divisions with the
tcp_ms_to_ticks() helper introduced in the initial RTO commits.

This unifies ceiling-division rounding for all timeout comparisons
(FIN_WAIT_2, SYN_RCVD, LAST_ACK, TIME_WAIT, keepalive, user timeout).

Signed-off-by: Tomer Cabouly <tcabouly@nvidia.com>

tomerdbz committed:
issue: 4043157 Set TCP initial RTO to 1s per RFC 6298

Set the initial TCP RTO to 1 second.
Add SYN retransmission backoff and fallback RTO behavior.
Use a shared timer conversion helper for TCP timeouts.

Signed-off-by: Tomer Cabouly <tcabouly@nvidia.com>

tomerdbz commented on src/core/util/sys_vars.cpp:
```diff
@@ -2744,6 +2754,16 @@ void mce_sys_var::configure_memory_limits(const config_registry &registry)
+                    "Clamping to 500ms to ensure protocol compliance.\n",
+                    SYS_VAR_TCP_TIMER_RESOLUTION_MSEC, tcp_timer_resolution_msec);
+        tcp_timer_resolution_msec = 500;
+    }
```
removed this from the PR

pasis commented on src/core/lwip/tcp_impl.h:
```diff
@@ -320,8 +320,57 @@ extern u32_t rcv_wnd_scale;
+
+static inline u32_t tcp_min_rto(void)
+{
+    return tcp_ms_to_rto_ticks(1000U);
```
1 second minimum RTO looks too high. Currently we have ~600ms RTO and 1 seconds will make recovery worse. We need to move towards shorter RTO, such as 200ms (currently problematic due to TCP timer granularity).

tomerdbz replied:
I learned something new today, linux violates RFC!

<img width="1893" height="862" alt="Image" src="https://github.com/user-attachments/assets/20a1f9dd-5fa0-420c-b46c-203eaf34992f" />

and you're right we should too! will fix

tomerdbz commented:
bot:retest

tomerdbz commented on src/core/lwip/tcp_impl.h:
```diff
@@ -320,8 +320,47 @@ extern u32_t rcv_wnd_scale;
+extern u32_t slow_tmr_interval;
 extern ip_route_mtu_fn external_ip_route_mtu;
 
+static inline u32_t tcp_ms_to_ticks(u32_t ms)
+{
+    return (u32_t)(((u64_t)ms + slow_tmr_interval - 1U) / slow_tmr_interval);
+}
```
can't be zero though
